### PR TITLE
Issue319 file path

### LINKED ISSUE #319: File path is needed in MMD files
The `storage_information` field should be populated when using the API as well. It can be added via the `ncheader2json` by using the function `netCDF4.Dataset.filepath`. The file size field is challenging to estimate for large files, since we do not want to copy them. We may need to change the MMD xsd file (`mmd_strict.xsd`) to make the file size not required, or create another workaround.

### PR DESCRIPTION
**Pull request by mortenwh**
**Summary**: closes #319 

**Related issue**: 

**Suggested reviewer(s)**:

**Reviewer checklist**:

- [ ] The headers of all files contain a reference to the repository license (i.e., "License: This file is part of py-mmd-tools, licensed under the Apache License 2.0 (https://www.apache.org/licenses/LICENSE-2.0)")
- [ ] 100% test coverage of new code - meaning:
    - [ ] The overall test coverage increased or remained the same as before
    - [ ] Every function is accompanied with a test suite
    - [ ] Tests are both positive (testing that the function work as intended with valid data) and negative (testing that the function behaves as expected with invalid data, e.g., that correct exceptions are thrown)
    - [ ] Functions with optional arguments have separate tests for all options
- [ ] Examples are supported by doctests
- [ ] All tests are passing
- [ ] All names (e.g., files, classes, functions, variables) are explicit
- [ ] Documentation (as docstrings) is complete and understandable

The checklist is based on the S-ENDA conventions and definition of done (see https://s-enda-documentation.readthedocs.io/en/latest/general_conventions.html). The above points are not necessarily relevant to all contributions. In that case, please add a short explanation to help the reviewer.

### CODE REVIEW

File: tests/test_nc_to_mmd.py [hunk 1]
```diff
@@ -532,6 +533,32 @@ def test_json_dry_run(dataDir):
...
+
+    with pytest.raises(ValueError) as ee:
+        tmp = Nc_to_mmd(test_json_header, json_input=True, file_size=10,
+                        opendap_url="https://thredds.met.no/etc", output_file="somefn.xml")
+    assert "Input parameter 'target_nc_filename' must be " in str(ee.value)
+
+
+@pytest.mark.py_mmd_tools
+def testNc_to_mmd_Init(dataDir):
+    """Additional tests for the init method.
```
Comment (Teddy-1000) [inline tests/test_nc_to_mmd.py:553]: test name and docstring should reflect what is aspect of init is being tested.

File: py_mmd_tools/nc_to_mmd.py [hunk 2]
```diff
@@ -262,8 +273,33 @@ def __init__(
...
+            self.file_size = pathlib.Path(self.netcdf_file).stat().st_size / (1024 * 1024)
+
+        else:
+            if file_size is None and not check_only:
+                raise ValueError("Input parameter 'file_size' must be "
+                                 "provided when using input from json.")
+            if target_nc_filename is None and not check_only:
+                raise ValueError("Input parameter 'target_nc_filename' must be "
+                                 "provided when using input from json.")
+            else:
```
Comment (Teddy-1000) [inline py_mmd_tools/nc_to_mmd.py:291]: Line 291 and 292 have duplicate functionality of line 300 and 301.

File: py_mmd_tools/nc_to_mmd.py [hunk 3]
```diff
@@ -262,8 +273,33 @@ def __init__(
             )
         super(Nc_to_mmd, self).__init__()
         self.output_file = output_file
+
         if not json_input:
             self.netcdf_file = os.path.abspath(netcdf_file)
+            self.target_nc_filename = self.netcdf_file
+            self.ncin = self.read_nc_file(self.netcdf_file)
+            self.check_attributes_not_empty(self.ncin)
+            self.file_size = pathlib.Path(self.netcdf_file).stat().st_size / (1024 * 1024)
```
Comment (Teddy-1000) [inline py_mmd_tools/nc_to_mmd.py:282]: This calculation is in conflict with docstring line 260. Bytes to GB is 1/1024**3

File: py_mmd_tools/nc_to_mmd.py [hunk 4]
```diff
@@ -1785,13 +1807,13 @@ def to_mmd(
...
             # Set storage_information
             self.metadata["storage_information"] = {
-                "file_name": os.path.basename(self.netcdf_file),
-                "file_location": os.path.dirname(self.netcdf_file),
+                "file_name": os.path.basename(self.target_nc_filename),
+                "file_location": os.path.dirname(self.target_nc_filename),
                 "file_format": "NetCDF-CF",
-                "file_size": "%.2f" % file_size,
+                "file_size": "%.2f" % self.file_size,
                 "file_size_unit": "MB",
```
Comment (Teddy-1000) [inline py_mmd_tools/nc_to_mmd.py:1817]: Also conflict with docstring.

File: py_mmd_tools/nc_to_mmd.py [hunk 5]
```diff
@@ -262,8 +273,33 @@ def __init__(
...
+                raise ValueError("Input parameter 'file_size' must be "
+                                 "provided when using input from json.")
+            if target_nc_filename is None and not check_only:
+                raise ValueError("Input parameter 'target_nc_filename' must be "
+                                 "provided when using input from json.")
+            else:
+                self.target_nc_filename = target_nc_filename
+            self.ncin = nc_wrapper(netcdf_file)
+            self.check_attributes_not_empty(self.ncin)
+            self.file_size = file_size
```
Comment (Teddy-1000) [inline py_mmd_tools/nc_to_mmd.py:295]: Conflict in documentation and code for unit of file size.

File: tests/test_nc_to_mmd.py [hunk 6]
```diff
@@ -532,6 +533,32 @@ def test_json_dry_run(dataDir):
...
+        mp.setattr(nc_wrapper, "getncattr", lambda *a, **k: ["what", "ever"])
+        with pytest.raises(ValueError) as ee:
+            tmp = Nc_to_mmd(test_json_header, json_input=True,
+                            opendap_url="https://thredds.met.no/etc", output_file="somefn.xml")
+        assert "Input parameter 'file_size' must be " in str(ee.value)
+
+    with pytest.raises(ValueError) as ee:
+        tmp = Nc_to_mmd(test_json_header, json_input=True, file_size=10,
+                        opendap_url="https://thredds.met.no/etc", output_file="somefn.xml")
+    assert "Input parameter 'target_nc_filename' must be " in str(ee.value)
```
Comment (Teddy-1000) [inline tests/test_nc_to_mmd.py:548]: Error message are not functional, I don't think updating error messages should break unittesting.

File: py_mmd_tools/nc_to_mmd.py (hunk 3 above)
Comment (mortenwh) [reply py_mmd_tools/nc_to_mmd.py:282]: set to MB in the docstring

File: tests/test_nc_to_mmd.py (hunk 6 above)
Comment (mortenwh) [reply tests/test_nc_to_mmd.py:548]: Not sure about this. First, I don't think it is very important, as it is easy to update the test. Second, I can't think of another alternative right away. Please suggest a change if you insist.

**Comment by mortenwh** [comment]
@Teddy-1000 - setter pris på om du kan følge opp denne

File: tests/test_nc_to_mmd.py (hunk 6 above)
Comment (Teddy-1000) [reply tests/test_nc_to_mmd.py:548]: I think it should be sufficient to just test that the correct error is raised. No need to check the content of the error.

**Comment by Teddy-1000** [comment]
> As discussed in the chat, we may want to calculate the file size in the `nc_wrapper`. Also, maybe we could calculate a `check_sum` also, then? We skipped that earlier because it takes a lot of time, but we could maybe calculate it from parts of the nc-file instead of the whole?

If you are going to calculate a partial checksum for the file, you will need to find a way to accurately communicate to the end user what part of the file is included in the checksum. Or else the checksum will be useless for the end user, and again there would be no need for it.

**Comment by mortenwh** [comment]
> > As discussed in the chat, we may want to calculate the file size in the `nc_wrapper`. Also, maybe we could calculate a `check_sum` also, then? We skipped that earlier because it takes a lot of time, but we could maybe calculate it from parts of the nc-file instead of the whole?
> 
> If you are going to calculate a partial checksum for the file, you will need to find a way to accurately communicate to the end user what part of the file is included in the checksum. Or else the checksum will be useless for the end user, and again there would be no need for it.

I see.. Ok, let's skip that for now and focus on the current issue #319 - I think I have answered your comments?

**Comment by mortenwh** [comment]
Det dukket opp en del lokale endringer her som jeg ikke har gjort ferdig. Mulig de skal inn her... Må sjekke

**Comment by mortenwh** [comment]
@Teddy-1000 - jeg hadde begynt på en forenkling som jeg hadde glemt å gjøre ferdig. Sorry... Siste versjon er hakket enklere og bør løse #319 for API'et. Dette gir ikke mulighet for å legge inn alternativ lagringsplass i MMD fremfor faktisk filnavn+mappe.

**Review by Teddy-1000 — APPROVED** [review]

**Pull request merged**